pkg/ceph/osd/health.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2017 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package osd
+
+import (
+	"time"
+
+	"github.com/rook/rook/pkg/ceph/client"
+	"github.com/rook/rook/pkg/clusterd"
+)
+
+var (
+	healthCheckInterval = 60 * time.Second
+
+	// waitThreshold is period before an action is taken
+	waitThreshold = 300 * time.Second
+)
+
+// Monitor defines OSD process monitoring
+type Monitor struct {
+	context *clusterd.Context
+	agent   *OsdAgent
+
+	// lastStatus keeps track of OSDs status
+	// key - Process ID; value: time of the status change.
+	lastStatus map[int]time.Time
+}
+
+// NewMonitor instantiates OSD monitoring
+func NewMonitor(context *clusterd.Context, agent *OsdAgent) *Monitor {
+	return &Monitor{context, agent, make(map[int]time.Time)}
+}
+
+// Run runs monitoring logic for osd processe's status at set intervals
+func (m *Monitor) Run(stopCh chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			logger.Infof("Stopping OSD monitoring.")
+			return
+
+		case <-time.After(healthCheckInterval):
+			logger.Debug("Checking osd processes status.")
+			err := m.procStatus()
+			if err != nil {
+				logger.Infof("Failed OSD status check: %s", err)
+			}
+		}
+	}
+}
+
+func (m *Monitor) procStatus() error {
+	osdDump, err := client.GetOSDDump(m.context, m.agent.cluster.Name)
+	if err != nil {
+		return err
+	}
+
+	// TODO: handling of OSDs that are not tracked by Agent or OSDDump
```

Comment:
Have to figure out how to deal with a discrepancy, if found, between OSDs tracked by Agent and those listed by OSDDump.

Reply:
as noted in my comment below, osddump will always have more than the agent (in a cluster with more than 1 osd pod), since osd dump is cluster wide and the agent is just for 1 pod.  I think it's safe to just go with the agent's list since it should be authoritative for the OSDs running inside this pod.

Reply:
In this case no change is needed here, it already acts in this way, I will just cleanup a bit. The status check loop is based on what the Agent is tracking.